pkg/version/version.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//
+// This module lets one tag built binaries with version metadata.
+//
+// Currently two pieces of metadata tracked/provided:
+//   - Version: version number, by convention one provided by 'git describe'
+//   - Build:   build id, by convention the git SHA1 the binary has been built from.
+//
+// To enable automatic versioning metadata for your binary, you need to
+//
+//   1) import this package
+//   2) add the linker flags to override the dummy package variables, for instance:
+//        LDFLAGS=-ldflags \
+//          "-X=github.com/intel/cri-resource-manager/pkg/version.Version=<version> \
+//           -X=github.com/intel/cri-resource-manager/pkg/version.Build=<build-id>"
+//
+// Note that further metadata can be trivially added in a similar fashion:
+//
+//   1) add the corresponding variables to this modules
+//   2) arrange the default values to be correctly overridden during linking
+//   3) add printing of the new metadata to PrintVersionInfo()
+//
+
+package version
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+)
+
+// Default values of variables we'll override with the linker.
+var (
+	// Version is our version as given by 'git describe'.
+	Version = "<If you see this, you ain't doin' it right, Jimbo...>"
+	// Build is the SHA1 of the repository we've been built from.
+	Build = "<If you see this, you ain't doin' it right, Jimbo...>"
+)
+
+// PrintVersionInfo prints version information about this binary.
+func PrintVersionInfo() {
+	fmt.Printf("%s version information:\n", filepath.Base(os.Args[0]))
```

Comment:
NIT: would it be better to use a variable for the name of the program, too, instead of relying on `os.Args[0]`? So that e.g. symlinked or renamed binaries would show the correct name. WDYT?

Reply:
You mean that we'd have something like version.Binary to represent the canonical binary name which we'd then set, using the linker, for instance to the basename of the linked binary ? That could be easily added in a similar fashion if it is deemed useful.

That said, it's a long un*x tradition to let symlinked or linked binaries imagine they have their own identity. Even so much that in some cases (compress, uncompress, gzip, etc.) the binary basename dictates default behaviour (in the aforementioned cases for instance typically implying -d on the command line).

Reply:
> That said, it's a long un*x tradition to let symlinked or linked binaries imagine they have their own identity.

Sure, it depends on the binary:
```
$ ln -s `which tar` foo
$ ./foo --version
tar (GNU tar) 1.30
Copyright (C) 2017 Free Software Foundation, Inc.
... 
$
$ ln -s `which gzip` bar
$ ./bar --version
bar 1.9
Copyright (C) 2017 Free Software Foundation, Inc.
Copyright (C) 1993 Jean-loup Gailly.
...
```

My thinking was that in our case (kind of "single-identity" binary) it could be more logical to function like `tar` above. But, I'm completely fine with `Args[0]`, too. We might have more interesting topics, too :smile: